### ISSUE TITLE
.Net: Keep context between Handlebars Calls

### ISSUE THREAD

**Issue by sanchez** [microsoft/semantic-kernel#4072]
Hi I have been reading through some of the code here but I can't seem to come up with a way of passing the context between multiple plans. I am currently using RC3, any help appreciated thanks!

My code currently looks like:
```csharp
var kernelSettings = KernelSettings.LoadSettings();

KernelBuilder builder = new();

builder.WithAzureOpenAIChatCompletion(
    kernelSettings.DeploymentId,
    kernelSettings.ModelId,
    kernelSettings.Endpoint,
    kernelSettings.ApiKey);

builder.WithPlugins(plugins =>
{
    plugins.AddPluginFromObject<Plugins.LightPlugin>();
    plugins.AddPluginFromObject<Plugins.Geometry.PointPlugin>();
});

Kernel kernel = builder.Build();

#pragma warning disable SKEXP0061 // Type is for evaluation purposes only and is subject to change or removal in future updates. Suppress this diagnostic to proceed.
#pragma warning disable SKEXP0060 // Type is for evaluation purposes only and is subject to change or removal in future updates. Suppress this diagnostic to proceed.

var planner = new HandlebarsPlanner(new HandlebarsPlannerConfig()
{
    AllowLoops = true
});

while (true)
{
    Console.ForegroundColor = ConsoleColor.DarkGreen;
    Console.Write("User > ");
    Console.ResetColor();
    string userRequest = Console.ReadLine()!;

    HandlebarsPlan plan = await planner.CreatePlanAsync(kernel, userRequest);

    string result = plan.Invoke(kernel, new KernelArguments());

    Console.ForegroundColor = ConsoleColor.DarkMagenta;
    Console.Write("Assistant > ");
    Console.ResetColor();
    Console.WriteLine(result.Trim());
}

#pragma warning restore SKEXP0060 // Type is for evaluation purposes only and is subject to change or removal in future updates. Suppress this diagnostic to proceed.
#pragma warning restore SKEXP0061 // Type is for evaluation purposes only and is subject to change or removal in future updates. Suppress this diagnostic to proceed.
```

The code inside the PointPlugin is two functions, one to create a (X, Y, Z) point in 3d space, and a function to calculate the length from (0, 0, 0) of said point.

**Comment by matthewbolanos** [comment]
To pass in context, you'll like need to create a string outside of your loop that you can use to store your overall context. During each loop of your while loop, you can then append to the string. You would then pass this in as part of the ask to the Handlebars planners.

As you make your app more advanced, you'll want to store the context in a vector memory provider, and then you could recall information at runtime to populate the correct context for the planner.

**Comment by matthewbolanos** [comment]
Reopening because it would be _better_ if the planners could take in a chat history object to preserve context from a conversation.

**Comment by matthewbolanos** [comment]
This will likely be fixed +V1.0 in the new year.

**Comment by sanchez** [comment]
> To pass in context, you'll like need to create a string outside of your loop that you can use to store your overall context. During each loop of your while loop, you can then append to the string. You would then pass this in as part of the ask to the Handlebars planners.
> 
> As you make your app more advanced, you'll want to store the context in a vector memory provider, and then you could recall information at runtime to populate the correct context for the planner.

Sorry just a quick question here, so a lot of what I am trying to test out right now is building on having native types being constructed through conversation. So there is no way I can keep the native types in a Dictionary and have handlebars be able to refer to them instead of requiring parsing of the entire chat history on each execution?

Cheers,
Sanchez

**Comment by teresaqhoang** [comment]
> Reopening because it would be _better_ if the planners could take in a chat history object to preserve context from a conversation.

@matthewbolanos We should create a separate issue for this as a new feature request + details on how the chat history should be used by the planner (i.e., additional context at the end?) 

Seems adjacent to feature request to inject custom domain knowledge https://github.com/microsoft/semantic-kernel/issues/4067 -- could we treat chat history as custom knowledge in its own partial?

**Comment by teresaqhoang** [comment]
> > To pass in context, you'll like need to create a string outside of your loop that you can use to store your overall context. During each loop of your while loop, you can then append to the string. You would then pass this in as part of the ask to the Handlebars planners.
> > As you make your app more advanced, you'll want to store the context in a vector memory provider, and then you could recall information at runtime to populate the correct context for the planner.
> 
> Sorry just a quick question here, so a lot of what I am trying to test out right now is building on having native types being constructed through conversation. So there is no way I can keep the native types in a Dictionary and have handlebars be able to refer to them instead of requiring parsing of the entire chat history on each execution?
> 
> Cheers, Sanchez

We have some work planned to split each section of the planner prompt into isolated partials that we'll structure together piece-by-piece, so users will have more control on the order / structure of the planner prompt (i.e., omit or add any custom sections of their choosing). You can track it here: https://github.com/microsoft/semantic-kernel/issues/4067

But for now, the only way you'd be able to add your native types are by concatenating it to the goal you pass to the planner. 

If you take a look at [CreatePlanPrompt.handlebars](https://github.com/microsoft/semantic-kernel/blob/240a2cbeaccc2e9562fa9aa82c856f74c5b5ace4/dotnet/src/Planners/Planners.Handlebars/Handlebars/CreatePlanPrompt.handlebars#L112C1-L113C1), this is how we inject complex type definitions.
![image](https://github.com/microsoft/semantic-kernel/assets/125500434/3309b431-bed6-43d8-83cb-cf1176d54189)

By default, all complex types are extracted from native functions registered on the kernel and injected into the planner prompt. 

If you want to inject other type definitions that aren't tied to a function, you can try an approach like this:

```c#
        // Create a set of arguments containing the native types definitions
        var arguments = new KernelArguments()
        {
            { "nativeTypeDefinitions", nativeTypesDictionary }
        };

        // Create custom prompt template and render using the Handlebars prompt template factory
        var customTypesPrompt = @"## Native types
... Some description of the native types, assuming they follow the `KernelParameterMetadata` definition

{{#each nativeTypeDefinitions}}
### {{Name}}:
{
  ""type"": ""object"",
  ""properties"": {
    {{#each Properties}}
    ""{{Name}}"": {
        ""type"": ""{{ParameterType.Name}}"",
    },
    {{/each}}
    }
}";
        var promptTemplateConfig = new PromptTemplateConfig()
        {
            Template = customTypesPrompt,
            TemplateFormat = HandlebarsPromptTemplateFactory.HandlebarsTemplateFormat,
            Name = "MyTypesDefinitionPrompt",
        };

        var handlebarsTemplate = new HandlebarsPromptTemplateFactory().Create(promptTemplateConfig);
        var definitionsCompiled = await handlebarsTemplate!.RenderAsync(kernel, arguments).ConfigureAwait(true);

        // Concatenate your native types definitions with the goal and pass it into the planner
        goal = $"Goal: {goal}\n\n {definitionsCompiled}";
        var plan = await planner.CreatePlanAsync(kernel, goal);
```

Sorry there's so many loops to jump through here, I'll work with @matthewbolanos on ways we can make the planner more customizable.

**Comment by matthewbolanos** [comment]
This should be addressed with partial support in #4067, closing as a dupe
